File: frontend/customer-portal/src/lib/types/transaction.ts
// Transaction Types
export interface Transaction {
  id: number;
  amount: number;
  merchant_name: string;
  transaction_type: TransactionType;
  description?: string;
  location?: string;
  device_info?: string;
  ip_address?: string;
  risk_score: number;
  classification: RiskClassification;
  status: TransactionStatus;
  risk_factors?: Record<string, any>;
  model_predictions?: Record<string, number>;
  explanation?: string;
  created_at: string;
  updated_at?: string;
  user_response?: 'YES' | 'NO';
  responded_at?: string;
}

export enum TransactionType {
  ONLINE = 'ONLINE',
  IN_STORE = 'IN_STORE',
  ATM = 'ATM',
  TRANSFER = 'TRANSFER'
}

export enum TransactionStatus {
  PENDING = 'PENDING',
  APPROVED = 'APPROVED',
  REJECTED = 'REJECTED',
  BLOCKED = 'BLOCKED'
}

export enum RiskClassification {
  SAFE = 'SAFE',
  SUSPICIOUS = 'SUSPICIOUS',
  FRAUD = 'FRAUD'
}

// Transaction Submit Request
export interface TransactionSubmitRequest {
  amount: number;
  merchant_name: string;
  transaction_type: TransactionType;
  description?: string;
  card_number: string;
  cardholder_name: string;
  cvv: string;
  expiry_date: string;
  billing_address?: string;
  location?: string;
  device_info?: string;
  ip_address?: string;
}

// Transaction Response
export interface TransactionResponse extends Transaction {}

// Transaction List Response
export interface TransactionListResponse {
  transactions: Transaction[];
  total: number;
  page: number;
  limit: number;
}

File: backend/models/transaction.py
from sqlalchemy import Column, Integer, String, Float, DateTime, Enum as SQLEnum, ForeignKey, JSON, Text
from sqlalchemy.sql import func
from database.connection import Base
import enum

class TransactionType(str, enum.Enum):
    ONLINE = "ONLINE"
    IN_STORE = "IN_STORE"
    ATM = "ATM"
    TRANSFER = "TRANSFER"

class TransactionStatus(str, enum.Enum):
    PENDING = "PENDING"
    APPROVED = "APPROVED"
    REJECTED = "REJECTED"
    BLOCKED = "BLOCKED"

class RiskClassification(str, enum.Enum):
    SAFE = "SAFE"
    SUSPICIOUS = "SUSPICIOUS"
    FRAUD = "FRAUD"

class Transaction(Base):
    __tablename__ = "transactions"
    
    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    amount = Column(Float, nullable=False)
    merchant_name = Column(String(200), nullable=False)
    transaction_type = Column(SQLEnum(TransactionType), nullable=False)
    description = Column(Text, nullable=True)
    
    risk_score = Column(Float, nullable=True)
    classification = Column(SQLEnum(RiskClassification), nullable=True)
    status = Column(SQLEnum(TransactionStatus), default=TransactionStatus.PENDING, nullable=False)
    user_response = Column(String(10), nullable=True)
    
    features = Column(JSON, nullable=True)
    model_predictions = Column(JSON, nullable=True)
    risk_factors = Column(JSON, nullable=True)
    
    location = Column(String(100), nullable=True)
    device_info = Column(String(200), nullable=True)
    ip_address = Column(String(45), nullable=True)
    
    # Card payment fields (encrypted)
    card_number_encrypted = Column(String(500), nullable=True)  # Encrypted card number
    cardholder_name = Column(String(200), nullable=True)
    cvv_encrypted = Column(String(500), nullable=True)  # Encrypted CVV
    expiry_date = Column(String(10), nullable=True)  # Format: MM/YY
    billing_address = Column(Text, nullable=True)
    
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    responded_at = Column(DateTime(timezone=True), nullable=True)
